# 4.1.
如图的二次四边形单元，试计算$\frac{\partial{N_1}}{\partial{x}}$和$\frac{\partial{N_2}}{\partial{y}}$
在自然坐标为$(1/2,1/2)$的点的数值 。

In [1]:
import sympy as sp
from sympy import Matrix, Rational
from sympy.abc import xi, eta, x, y
from finite_element.interpolation import Serendipity, SurfaceExclude, Interpolation, Isoparametric
sp.init_printing()


## 计算插值函数

In [10]:
surface_exclude = SurfaceExclude(axises=[xi, eta])
surfaces = [xi+1, xi-1, eta+1, eta-1]
local_coordinates = [
    [1, 1], [-1, 1], [-1, -1], [1, -1],
    [0, 1], [-1, 0], [0, -1], [1, 0],
]
[surface_exclude.add_surface(surface) for surface in surfaces]
interpolation_functions = [
    surface_exclude.get_interpolation_function(local_coordinate)
    for local_coordinate in local_coordinates]
Matrix(interpolation_functions)


⎡     (η + 1)⋅(ξ + 1)     ⎤
⎢     ───────────────     ⎥
⎢            4            ⎥
⎢                         ⎥
⎢    -(η + 1)⋅(ξ - 1)     ⎥
⎢    ─────────────────    ⎥
⎢            4            ⎥
⎢                         ⎥
⎢     (η - 1)⋅(ξ - 1)     ⎥
⎢     ───────────────     ⎥
⎢            4            ⎥
⎢                         ⎥
⎢    -(η - 1)⋅(ξ + 1)     ⎥
⎢    ─────────────────    ⎥
⎢            4            ⎥
⎢                         ⎥
⎢-(η + 1)⋅(ξ - 1)⋅(ξ + 1) ⎥
⎢─────────────────────────⎥
⎢            2            ⎥
⎢                         ⎥
⎢ (η - 1)⋅(η + 1)⋅(ξ - 1) ⎥
⎢ ─────────────────────── ⎥
⎢            2            ⎥
⎢                         ⎥
⎢ (η - 1)⋅(ξ - 1)⋅(ξ + 1) ⎥
⎢ ─────────────────────── ⎥
⎢            2            ⎥
⎢                         ⎥
⎢-(η - 1)⋅(η + 1)⋅(ξ + 1) ⎥
⎢─────────────────────────⎥
⎣            2            ⎦

## 计算等参元的雅克比矩阵

整体坐标下的各点位坐标

In [13]:
global_coordinates = [
    [40, 50], [5, 40], [10, 10], [30, 20],
    [Rational(45,2), Rational(90,2)],
    [Rational(15,2), Rational(50,2)],
    [Rational(40,2), Rational(30,2)],
    [Rational(70,2), Rational(70,2)],
]
Matrix(global_coordinates)

⎡ 40   50⎤
⎢        ⎥
⎢ 5    40⎥
⎢        ⎥
⎢ 10   10⎥
⎢        ⎥
⎢ 30   20⎥
⎢        ⎥
⎢45/2  45⎥
⎢        ⎥
⎢15/2  25⎥
⎢        ⎥
⎢ 20   15⎥
⎢        ⎥
⎣ 35   35⎦

In [14]:
isoparametric = Isoparametric(
    local_axises=[xi, eta],
    global_axises=[x, y],
    interpolation_functions=interpolation_functions,
    coordinates=global_coordinates)
isoparametric.jacobian

⎡      2                                                        ⎤
⎢  55⋅η    5⋅η⋅ξ   15⋅η   85⋅ξ   55       2                     ⎥
⎢- ───── - ───── + ──── - ──── + ──  - 5⋅η  - 30⋅η⋅ξ - 60⋅ξ + 10⎥
⎢    4       2      4      2     2                              ⎥
⎢                                                               ⎥
⎢                     2                                         ⎥
⎢  55⋅η⋅ξ   85⋅η   5⋅ξ    15⋅ξ   5                        2     ⎥
⎢- ────── - ──── - ──── + ──── + ─   -10⋅η⋅ξ - 60⋅η - 15⋅ξ  + 30⎥
⎣    2       2      4      4     2                              ⎦

## 计算等参元对整体坐标的偏导数

In [15]:
isoparametric.displacement_interpolation_function_diff_global

⎡                 ⎛ 2         2        2                    2           ⎞     
⎢                -⎝η ⋅ξ + 11⋅η  - 3⋅η⋅ξ  - 4⋅η⋅ξ + 6⋅η - 9⋅ξ  - 10⋅ξ - 4⎠     
⎢─────────────────────────────────────────────────────────────────────────────
⎢  ⎛    3       2  2        2          2          2               3       2   
⎢5⋅⎝98⋅η  - 96⋅η ⋅ξ  - 183⋅η ⋅ξ - 100⋅η  - 187⋅η⋅ξ  - 178⋅η + 90⋅ξ  - 28⋅ξ  - 
⎢                                                                             
⎢                    2         2      2                       2               
⎢                11⋅η ⋅ξ + 23⋅η  - η⋅ξ  + 20⋅η⋅ξ + 35⋅η - 33⋅ξ  - 15⋅ξ + 20   
⎢─────────────────────────────────────────────────────────────────────────────
⎢   ⎛    3       2  2        2          2          2               3       2  
⎣20⋅⎝98⋅η  - 96⋅η ⋅ξ  - 183⋅η ⋅ξ - 100⋅η  - 187⋅η⋅ξ  - 178⋅η + 90⋅ξ  - 28⋅ξ  -

                                  2         2        2                    2   
                                 η ⋅ξ + 13⋅η  - 3⋅η

代入求值

In [16]:
isoparametric.displacement_interpolation_function_diff_global.subs({xi: Rational(1, 2), eta: Rational(1, 2)})

⎡       -19           37   -11    31     59          ⎤
⎢-1/85  ────  1/255  ────  ────  ────   ────    1/30 ⎥
⎢       2295         2295  510   4590   4590         ⎥
⎢                                                    ⎥
⎢       109          -19    47    257    193         ⎥
⎢-1/68  ────  1/204  ────  ────  ─────  ─────  -1/120⎥
⎣       9180         9180  2040  18360  18360        ⎦

其中，整体坐标下，$\frac{\partial{N_1}}{\partial{x}}=-\frac{1}{85}$和$\frac{\partial{N_2}}{\partial{y}}=\frac{109}{9180}$

库代码见[这里](https://github.com/panhaoyu/finite_element)